Update error guidance for Azure data plane services (#311)
Microsoft/api-guidelines: azure/Guidelines.md

:white_check_mark: **DO** return an `x-ms-error-code` response header with a string error code indicating what went wrong.

> *NOTE: `x-ms-error-code` values are part of your API contract (because customer code is likely to do comparisons against them) and cannot change in the future.*

Reviewer: ```suggestion
*NOTE: `x-ms-error-code` values are part of your API contract (because customer code is likely to do comparisons against them) and cannot change in the future without a new version of your API*
```

Reviewer 2: You cannot change them even if there is a new api-version. Doing so is breaking.

> :heavy_check_mark: **YOU MAY** implement the `x-ms-error-code` values as an enum with `"modelAsString": true` because it's possible add new values over time.  In particular, it's only a breaking change if the same conditions result in a different top-level error code.

Reviewer: ```suggestion
:heavy_check_mark: **YOU MAY** implement the `x-ms-error-code` values as an [Extensible Enum](#use-extensible-enums)--an enum with `"modelAsString": true`--because it's possible add new values over time.  In particular, it's only a breaking change if the same conditions result in a *different* top-level error code. 
```

Reviewer 2: I think it would be nice to separate the error concepts from the swagger suggestions. The service implementors need to know what's in the contract and how to craft an error. This is not related to swagger.
The swagger is how to describe the implementation and explaining these rules is useful but I think should be separated out because different people are likely to care about these 2 things differently.

Author: That's a fair suggestion, but I'd still say if the service designers aren't thinking about how they'll eventually do this in the swagger then we'll be told it's an impossible amount of work to pull the list of errors out of the service's front end.  Would you separate this out into a separate guideline?  Or remove from the document entirely?

Reviewer: @JeffreyRichter -- We already talk about `Extensible Enums` in the spec, see: [Extensible Enums](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#use-extensible-enums). These are referenced in a number of places. We did this, in part, because at the time, we didn't have the [Open API Style Guidelines](https://aka.ms/azapi/style). What I'd like to do is be consistent within the Guidelines that we have now, and point to the extensible enum section. We will open a new issue to determine if we should refactor the guidelines to remove the section to the style guide--which we can do on a separate pass. Captured this thought in a new issue: #316.

Property | Type | Required | Description
-------- | ---- | :------: | -----------

> `error` | ErrorDetail | ✔ | The top-level error object whose `code` matches `x-ms-error-code`

Reviewer 3: `x-ms-error-code` response header

:heavy_check_mark: **YOU MAY** treat the other fields as you wish as they are _not_ considered part of your service's API contract and customers should not take a dependency on them or their value. They exist to help customers self-diagnose issues.

:heavy_check_mark: **YOU MAY** add additional properties for any data values in your error message so customers don't resort to parsing your error message.  For example, an error with `"message": "A maximum of 16 keys are allowed per account."` might also add a `"maximumKeys": 16` property.  This is not part of your API contract and should only be used for diagnosing problems.

*Note: Do not use this mechanism to provide information developers need to rely on in code (ex: the error message can give details about why you've been throttled, but the `Retry-After` should be what developers rely on to back off).*

> :warning: **YOU SHOULD NOT** use your OpenAPI/Swagger specification to document every failing status code or error code for each operation.

Reviewer 4: I think we want this to be a bit stronger.
```suggestion
:warning: **YOU SHOULD NOT** document specific error status codes in your OpenAPI/Swagger spec unless the "default" response cannot properly describe the specific error response (e.g. body schema is different).
```
